File: apps/android/app/src/test/java/com/exponential/app/domain/PrGraphTest.kt
package com.exponential.app.domain

import com.exponential.app.data.db.CodingSessionEntity
import com.exponential.app.data.db.IssueEntity
import com.exponential.app.ui.session.sessionRowTitle
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

// EXP-897 part 4 — the badge's model, the same four tests web
// (`pr-graph.test.ts`), iOS (`PrGraphTests`) and the desktop (`pr_graph`) run.
class PrGraphTest {

    private fun issue(
        id: String,
        identifier: String = id.uppercase(),
        branch: String? = null,
        base: String? = null,
        prUrl: String? = null,
    ) = IssueEntity(
        id = id,
        boardId = "board-1",
        number = 1,
        identifier = identifier,
        title = "Issue $identifier",
        status = "backlog",
        priority = "none",
        sortOrder = 1.0,
        prUrl = prUrl,
        branch = branch,
        prBaseBranch = base,
        createdAt = "2026-09-10T10:00:00Z",
        updatedAt = "2026-09-10T10:00:00Z",
    )

    private fun session(
        id: String,
        parent: String? = null,
        issueId: String? = null,
        // EXP-876: what makes a run a BATCH, and what names it.
        batchIssueIds: String? = null,
        branch: String? = null,
        actionName: String? = null,
    ) =
        CodingSessionEntity(
            id = id,
            issueId = issueId,
            teamId = "team-1",
            userId = "user-1",
            branch = branch,
            batchIssueIds = batchIssueIds,
            actionName = actionName,
            parentSessionId = parent,
            startedAt = "2026-09-10T10:00:00Z",
            createdAt = "2026-09-10T10:00:00Z",
            updatedAt = "2026-09-10T10:00:00Z",
        )

    private fun graph(
        issue: IssueEntity?,
        issues: List<IssueEntity>,
        session: CodingSessionEntity? = null,
        sessions: List<CodingSessionEntity> = emptyList(),
    ) = PrGraph.build(issue, session, issues, sessions, emptyList())

    @Test
    fun reportsAStackBadgeForAStackedPr() {
        val bottom = issue("a", branch = "exp/A")
        val top = issue("b", branch = "exp/B", base = "exp/A")
        val built = graph(top, listOf(bottom, top))

        assertEquals(PrGraph.BadgeKind.STACK, PrGraph.badgeKind(built))
        assertEquals(listOf("a", "b"), built.stack.map { it.entry.representative.id })
        assertEquals(listOf(0, 1), built.stack.map { it.depth })
        assertEquals(1, built.subject?.depth)
        assertNull(built.batch)
    }

    @Test
    fun reportsABatchBadgeForABatchPr() {
        val first = issue("a", prUrl = "https://github.com/o/r/pull/7", branch = "exp/batch-1")
        val second = issue("b", prUrl = "https://github.com/o/r/pull/7", branch = "exp/batch-1")
        val built = graph(first, listOf(first, second))

        assertEquals(PrGraph.BadgeKind.BATCH, PrGraph.badgeKind(built))
        assertEquals(listOf("a", "b"), built.batch?.issues?.map { it.id })
        assertEquals(1, built.stack.size)
    }

    @Test
    fun reportsBothForABatchInsideAStack() {
        val foundation = issue("f", branch = "exp/F")
        val first = issue("a", prUrl = "https://github.com/o/r/pull/7", branch = "exp/batch-1", base = "exp/F")
        val second = issue("b", prUrl = "https://github.com/o/r/pull/7", branch = "exp/batch-1", base = "exp/F")
        val built = graph(first, listOf(foundation, first, second))

        assertEquals(PrGraph.BadgeKind.STACK_AND_BATCH, PrGraph.badgeKind(built))
        // ONE node per pull request: the batch is a single stack member.
        assertEquals(2, built.stack.size)
        assertTrue(built.stack.last().entry.isBatch)
        assertEquals(listOf("a", "b"), built.batch?.issues?.map { it.id })
    }

    @Test
    fun reportsNothingForALonePr() {
        val lone = issue("a", branch = "exp/A", prUrl = "https://github.com/o/r/pull/3")
        val built = graph(lone, listOf(lone, issue("other", branch = "exp/O")))

        assertNull(PrGraph.badgeKind(built))
        assertEquals(1, built.stack.size)
        assertNull(built.batch)
    }

    // EXP-876: the pill and its sheet are the surface built to name work that
    // spans several issues — and a batch RUN, which spans them, resolved
    // nothing at all before this (it links no issue and stamps no pr_url).
    // Mirrored ×4.
    @Test
    fun reportsABatchBadgeForABatchRunBeforeItsPr() {
        val one = issue("one")
        val two = issue("two")
        val run = session("run", batchIssueIds = """["one","two"]""")
        val built = graph(null, listOf(one, two), session = run, sessions = listOf(run))

        assertEquals(PrGraph.BadgeKind.BATCH, PrGraph.badgeKind(built))
        // The composer's order, so the sheet reads like the row that named it.
        assertEquals(listOf("one", "two"), built.batch?.issues?.map { it.id })
        // No pull request yet: a batch of two is not a stack of two.
        assertEquals(1, built.stack.size)
    }

    @Test
    fun reportsABatchRunsPullRequestOnceItOpens() {
        // The GROUPED entry wins the moment the PR exists: it carries the
        // branch and the base the stack chains on, so a stacked batch still
        // reads `stack+batch` from its run.
        val url = "https://github.com/o/r/pull/9"
        val lower = issue("lower", branch = "exp/LOWER")
        val one = issue("one", branch = "exp/batch-abcd1234", base = "exp/LOWER", prUrl = url)
        val two = issue("two", branch = "exp/batch-abcd1234", base = "exp/LOWER", prUrl = url)
        val run = session(
            "run",
            batchIssueIds = """["one","two"]""",
            branch = "exp/batch-abcd1234",
        )
        val built = graph(null, listOf(lower, one, two), session = run, sessions = listOf(run))

        assertEquals(PrGraph.BadgeKind.STACK_AND_BATCH, PrGraph.badgeKind(built))
        assertEquals(2, built.batch?.issues?.size)
    }

    @Test
    fun reportsNothingForABatchRunWhoseIssuesAreUnknown() {
        // No stored ids and no PR: the row reads "Batch run" and wears no
        // pill, rather than a pill that could say nothing.
        val bare = session("run")
        val built = graph(null, emptyList(), session = bare, sessions = listOf(bare))
        assertNull(PrGraph.badgeKind(built))
        assertNull(built.batch)

        // An ACTION run is never a batch, whatever else it carries.
        val action = session("chat", actionName = "Chat", batchIssueIds = """["one"]""")
        val actionGraph =
            graph(null, listOf(issue("one")), session = action, sessions = listOf(action))
        assertNull(actionGraph.batch)
    }

    // EXP-930: what the OVERLAY draws on a batch's Run face — the covered
    // issues above the run tree, and every tree row named off the synced
    // rows. The sheet used to join no issue at all, so a batch whose issues
    // were already in the store still read `Issue syncing…`.
    @Test
    fun namesABatchRunsIssuesAndItsRunRows() {
        val one = issue("one")
        val two = issue("two")
        val pool = listOf(one, two)
        val batch = session("run", batchIssueIds = """["one","two"]""")
        val child = session("child", parent = "run", issueId = "two")
        val built = graph(null, pool, session = batch, sessions = listOf(batch, child))

        // The `Issues` section the sheet now draws above `Runs`.
        assertEquals(listOf("ONE", "TWO"), built.batch?.issues?.map { it.identifier })

        // EXP-968: the graph NAMES its own rows — the sheet renders these.
        assertEquals(listOf("Issue ONE", "Issue TWO"), built.tree.map { it.title })
        assertEquals(listOf(null, "two"), built.tree.map { it.issue?.id })
        // The bug itself: no joined issue, no name. ONE string ×4.
        assertEquals(ISSUE_SYNCING_TITLE, sessionRowTitle(child, null))
        assertEquals("Issue syncing…", ISSUE_SYNCING_TITLE)
    }

    // EXP-968: the Runs list of the stack overlay said "Issue not synced yet"
    // about every row, because the labels were joined against a SECOND issue
    // snapshot. A run tree whose issues are all synced names every row.
    @Test
    fun namesEveryRunRowFromTheGraphsOwnIssues() {
        val one = issue("one")
        val two = issue("two")
        val root = session("root", issueId = "one")
        val child = session("child", parent = "root", issueId = "two")
        val orphan = session("orphan", parent = "root", issueId = "missing")
        val built = graph(
            one,
            listOf(one, two),
            session = root,
            sessions = listOf(root, child, orphan),
        )

        assertEquals(listOf("root", "child", "orphan"), built.tree.map { it.session.id })
        assertEquals(
            listOf("Issue ONE", "Issue TWO", ISSUE_SYNCING_TITLE),
            built.tree.map { it.title },
        )
        // Only the row whose issue is genuinely missing waits on the sync.
        assertEquals(listOf("one", "two", null), built.tree.map { it.issue?.id })
    }

    @Test
    fun nestsTheRunFamilyOfTheShownSession() {
        val root = session("root")
        val child = session("child", parent = "root")
        val grand = session("grand", parent = "child")
        val stranger = session("stranger")
        val built = graph(
            null,
            emptyList(),
            session = child,
            sessions = listOf(stranger, grand, child, root),
        )
        assertEquals(listOf("root", "child", "grand"), built.tree.map { it.session.id })
        assertEquals(listOf(0, 1, 2), built.tree.map { it.depth })
    }
}

File: apps/web/src/lib/pr-graph.test.ts
import { describe, expect, it } from "vitest"
import { badgeChip, badgeKind, badgeShape, prGraph } from "./pr-graph"

// EXP-897 Part 4 — the badge model. Every `it` name here is mirrored by iOS
// PrGraphTests, Android PrGraphTest and the desktop `pr_graph` tests.

const issue = (
  id: string,
  over: Partial<{
    branch: string | null
    prBaseBranch: string | null
    prUrl: string | null
    status: string
    title: string
    createdAt: string
  }> = {}
) => ({
  id,
  identifier: id.toUpperCase(),
  status: over.status ?? `in_progress`,
  branch: over.branch ?? `exp/${id.toUpperCase()}`,
  prBaseBranch: over.prBaseBranch ?? null,
  prUrl: over.prUrl ?? `https://github.com/acme/app/pull/${id}`,
  // EXP-876: a batch run names itself off these.
  title: over.title ?? id.toUpperCase(),
  createdAt: over.createdAt ?? `2026-09-10T10:00:00Z`,
})

const session = (
  id: string,
  parentSessionId: string | null = null,
  issueId: string | null = null,
  over: Partial<{
    actionName: string | null
    batchIssueIds: string[] | null
    branch: string | null
  }> = {}
) => ({
  id,
  parentSessionId,
  startedAt: `2026-09-10T10:00:00Z`,
  issueId,
  actionName: over.actionName ?? null,
  batchIssueIds: over.batchIssueIds ?? null,
  branch: over.branch ?? null,
})

describe(`prGraph`, () => {
  it(`reports a stack badge for a stacked pr`, () => {
    const lower = issue(`lower`)
    const upper = issue(`upper`, { prBaseBranch: `exp/LOWER` })
    const graph = prGraph({
      issue: upper,
      issues: [lower, upper],
      sessions: [],
    })
    expect(badgeKind(graph)).toBe(`stack`)
    expect(graph.stack.map((row) => row.entry.issue.id)).toEqual([
      `lower`,
      `upper`,
    ])
    expect(graph.stack.map((row) => row.depth)).toEqual([0, 1])
    expect(graph.batch).toBeNull()
  })

  it(`reports a batch badge for a batch pr`, () => {
    const url = `https://github.com/acme/app/pull/9`
    const one = issue(`one`, { prUrl: url, branch: `exp/batch-abcd1234` })
    const two = issue(`two`, { prUrl: url, branch: `exp/batch-abcd1234` })
    const graph = prGraph({ issue: one, issues: [one, two], sessions: [] })
    expect(badgeKind(graph)).toBe(`batch`)
    expect(graph.batch?.issues.map((row) => row.id)).toEqual([`one`, `two`])
    expect(graph.stack).toEqual([])
  })

  it(`reports both for a batch inside a stack`, () => {
    const url = `https://github.com/acme/app/pull/9`
    const lower = issue(`lower`)
    const one = issue(`one`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    const two = issue(`two`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    const graph = prGraph({
      issue: two,
      issues: [lower, one, two],
      sessions: [],
    })
    expect(badgeKind(graph)).toBe(`stack+batch`)
    // The batch is ONE stack member, never two rows.
    expect(graph.stack.map((row) => row.entry.key)).toEqual([
      lower.prUrl,
      url,
    ])
    expect(graph.batch?.issues).toHaveLength(2)
  })

  // EXP-876: the pill and its sheet are the surface built to name work that
  // spans several issues — and a batch RUN, which spans them, resolved
  // nothing at all before this (it links no issue and stamps no pr_url).
  // Mirrored ×4.
  it(`reports a batch badge for a batch run before its pr`, () => {
    const one = issue(`one`, { prUrl: null })
    const two = issue(`two`, { prUrl: null })
    const run = session(`run`, null, null, { batchIssueIds: [`one`, `two`] })
    const graph = prGraph({ session: run, issues: [one, two], sessions: [run] })
    expect(badgeKind(graph)).toBe(`batch`)
    // The composer's order, so the sheet reads like the row that named it.
    expect(graph.batch?.issues.map((row) => row.id)).toEqual([`one`, `two`])
    // No pull request yet, so no stack — a batch of two is not a stack of two.
    expect(graph.stack).toEqual([])
  })

  it(`reports a batch run's pull request once it opens`, () => {
    // The GROUPED entry wins the moment the PR exists: it carries the branch
    // and the base the stack chains on, so a stacked batch still reads
    // `stack+batch` from its run.
    const url = `https://github.com/acme/app/pull/9`
    const lower = issue(`lower`)
    const one = issue(`one`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    const two = issue(`two`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    const run = session(`run`, null, null, {
      batchIssueIds: [`one`, `two`],
      branch: `exp/batch-abcd1234`,
    })
    const graph = prGraph({
      session: run,
      issues: [lower, one, two],
      sessions: [run],
    })
    expect(badgeKind(graph)).toBe(`stack+batch`)
    expect(graph.entry?.key).toBe(url)
    expect(graph.batch?.issues).toHaveLength(2)
  })

  it(`reports nothing for a batch run whose issues are unknown`, () => {
    // No stored ids and no PR: the row reads "Batch run" and wears no pill,
    // rather than a pill that could say nothing.
    const run = session(`run`)
    const graph = prGraph({ session: run, issues: [], sessions: [run] })
    expect(badgeKind(graph)).toBeNull()
    expect(graph.entry).toBeNull()
    // And an ACTION run is never a batch, whatever else it carries.
    const action = session(`action`, null, null, {
      actionName: `Chat`,
      batchIssueIds: [`one`],
    })
    expect(
      prGraph({ session: action, issues: [issue(`one`)], sessions: [action] })
        .batch
    ).toBeNull()
  })

  it(`reports nothing for a lone pr`, () => {
    const lone = issue(`lone`)
    const graph = prGraph({ issue: lone, issues: [lone], sessions: [] })
    expect(badgeKind(graph)).toBeNull()
  })

  // EXP-1079: the desktop's `is_visible` + `badge_glyphs` — a run with a
  // family earns the pill on the Run face alone, and a PR relation always
  // wins over it.
  it(`shapes a runs badge for a run with a family, on the run face only`, () => {
    const sessions = [session(`child`, `root`), session(`root`)]
    const family = prGraph({ session: sessions[0], issues: [], sessions })
    expect(badgeShape(family, `run`)).toBe(`runs`)
    expect(badgeShape(family, `changes`)).toBeNull()
    expect(badgeShape(family, `issue`)).toBeNull()
    const alone = prGraph({
      session: session(`alone`),
      issues: [],
      sessions: [session(`alone`)],
    })
    expect(badgeShape(alone, `run`)).toBeNull()
    const batchA = issue(`bata`, { prUrl: `https://github.com/acme/app/pull/9` })
    const batchB = issue(`batb`, { prUrl: `https://github.com/acme/app/pull/9` })
    const batched = prGraph({
      session: sessions[0],
      issue: batchA,
      issues: [batchA, batchB],
      sessions,
    })
    expect(badgeShape(batched, `run`)).toBe(`batch`)
  })

  // EXP-1058: the header's stacked chip — front issue + how many behind.
  // Mirrored ×4 (`badge_chip_*` desktop, PrGraphTests, PrGraphTest).
  it(`names the representative issue and the count on the stacked chip`, () => {
    const url = `https://github.com/acme/app/pull/9`
    const lower = issue(`lower`)
    const one = issue(`one`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    const two = issue(`two`, {
      prUrl: url,
      branch: `exp/batch-abcd1234`,
      prBaseBranch: `exp/LOWER`,
    })
    // A batch inside a stack: the subject PR's representative, every other
    // issue on the stack behind it.
    const both = prGraph({ issue: two, issues: [lower, one, two], sessions: [] })
    expect(badgeChip(both, `issue`)?.issue?.id).toBe(`one`)
    expect(badgeChip(both, `issue`)?.count).toBe(2)
    // A plain batch: the others of the batch.
    const batch = prGraph({ issue: one, issues: [one, { ...two, prBaseBranch: null }], sessions: [] })
    expect(badgeChip(batch, `changes`)?.count).toBe(1)
    // A run family with no issue: no front issue, the other runs behind.
    const sessions = [session(`child`, `root`), session(`root`)]
    const family = prGraph({ session: sessions[0], issues: [], sessions })
    expect(badgeChip(family, `run`)).toEqual({ issue: null, count: 1 })
    // No badge = no chip.
    expect(badgeChip(family, `issue`)).toBeNull()
  })

  it(`nests the subject run's whole tree, from its root`, () => {
    const sessions = [
      session(`child`, `root`),
      session(`root`),
      session(`stranger`),
    ]
    const graph = prGraph({
      session: sessions[0],
      issues: [],
      sessions,
    })
    expect(graph.tree.map((row) => `${row.session.id}@${row.depth}`)).toEqual([
      `root@0`,
      `child@1`,
    ])
  })

  it(`lists the subject issue's open blockers`, () => {
    const me = issue(`me`)
    const blocker = issue(`blocker`)
    const closed = issue(`closed`, { status: `done` })
    const graph = prGraph({
      issue: me,
      issues: [me, blocker, closed],
      sessions: [],
      relations: [
        { type: `blocks`, issueId: `blocker`, relatedIssueId: `me` },
        { type: `blocks`, issueId: `closed`, relatedIssueId: `me` },
      ],
    })
    expect(graph.blockedBy.map((row) => row.id)).toEqual([`blocker`])
  })
})
